lib/highline/terminal/unix_stty.rb

```diff
@@ -0,0 +1,91 @@
+class HighLine
+  class Terminal::UnixStty < Terminal
+
+    # A Unix savvy method using stty to fetch the console columns, and rows.
+    # ... stty does not work in JRuby
+    def terminal_size
+      begin
+        require "io/console"
+        winsize = IO.console.winsize.reverse rescue nil
+        return winsize if winsize
+      rescue LoadError
+      end
+
+      if /solaris/ =~ RUBY_PLATFORM and
+        `stty` =~ /\brows = (\d+).*\bcolumns = (\d+)/
+        [$2, $1].map { |c| x.to_i }
+      elsif `stty size` =~ /^(\d+)\s(\d+)$/
+        [$2.to_i, $1.to_i]
+      else
+        [ 80, 24 ]
+      end
+    end
+
+                              # *WARNING*:  This requires the external "stty" program!
+    CHARACTER_MODE = "unix_stty"   # For Debugging purposes only.
+
+    def raw_no_echo_mode
+      @state = `stty -g`
+      system "stty raw -echo -icanon isig"
```

Comment:
I've had to break compatibility for some time. And yes, all my tests are
running against this Terminal::UnixStty
The others are just ghosts of the older code.
Later, it will be much easier to add compatibily to other "terminals" like
Windows, JRuby.
If you look at the code in master, it relied on a lot of tests and
conditionals to know what to do. If JRuby, if windows and so... And these
conditionals were scattered throughout the code, making it hard to maintain.
I have tried to change conditionals with polymorphism.
When the code get ready, if you want to add a compatibility code you will
only inherit a class from HighLine::Terminal and it will work!

Thank you very much for this feedback and for any feedback to come. I
really appreciate.

Em Sáb, 13 de jun de 2015 16:24, Dāvis <notifications@github.com> escreveu:

> In lib/highline/terminal/unix_stty.rb
> <https://github.com/JEG2/highline/pull/140#discussion_r32373076>:
>
> > +      if /solaris/ =~ RUBY_PLATFORM and
> > +        `stty` =~ /\brows = (\d+).*\bcolumns = (\d+)/
> > +        [$2, $1].map { |c| x.to_i }
> > +      elsif `stty size` =~ /^(\d+)\s(\d+)$/
> > +        [$2.to_i, $1.to_i]
> > +      else
> > +        [ 80, 24 ]
> > +      end
> > +    end
> > +
> > +                              # *WARNING*:  This requires the external "stty" program!
> > +    CHARACTER_MODE = "unix_stty"   # For Debugging purposes only.
> > +
> > +    def raw_no_echo_mode
> > +      @state = `stty -g`
> > +      system "stty raw -echo -icanon isig"
>
> I don't know much about stty and how different implementations exist, but
> currently seems you could just include SystemExtensions::Stty because
> it's exactly same for raw_no_echo_mode and restore_mode methods.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JEG2/highline/pull/140/files#r32373076>.
>